File: src/main/scala/com/dadrox/sbt/junit/JunitLogger.scala
package com.dadrox.sbt.junit

import org.scalatools.testing.Logger

case class Color(value: String) {
    override val toString = value
}

object Colors {
    val Normal = Color("\u001b[m")
    val Green = Color("\u001b[32m")
    val Red = Color("\u001b[31m")
    val Yellow = Color("\u001b[33m")
    val Cyan = Color("\u001b[36m")
    val Grey = Color("\u001b[37m")
    val Magenta = Color("\u001b[35m")

    object Bright {
        val Green = Color("\u001b[1;32m")
        val Cyan = Color("\u001b[1;36m")
    }
}

case class FormattedMessage(msg: String)

class JunitLogger(
        loggers: Array[Logger],
        config: Config) {
    import Colors._

    val colorsEnabled = config.colorsEnabled

    //    def ansiCodesSupported(): Boolean = true

    def error(msg: => FormattedMessage) = loggers.foreach(_.error(msg.msg))
    def warn(msg: => FormattedMessage) = loggers.foreach(_.warn(msg.msg))
    def info(msg: => FormattedMessage) = loggers.foreach(_.info(msg.msg))
    def debug(msg: => FormattedMessage) = loggers.foreach(_.debug(msg.msg))
    def trace(t: => Throwable) = loggers.foreach(_.trace(t))

    def format(c: Color, msg: String) = FormattedMessage(colorsEnabled match {
        case true  => c + msg + Normal
        case false => msg.replaceAll("\\u001b.*m", "") // this may not work?
    })

    def skipped(msg: String) = warn(format(Yellow, msg))

    def testStarted(name: TestName) = {
        val msg = format(Cyan, "Started " + name.decorated)
        config.testStart match {
            case true  => info(msg)
            case false => debug(msg)
        }
    }
    def testPassed(event: Event) = {
        val msg = format(Green, "Passed " + event.testName.decorated)
        config.testPass match {
            case true  => info(msg)
            case false => debug(msg)
        }
    }
    def testFailed(name: TestFailure) = error(format(Magenta, "Failed " + name.decorated + ": " + name.failure.getMessage()))

    def suiteStarted(name: SuiteName) = {
        val msg = format(Bright.Cyan, "Started " + name.decorated)
        config.suiteStart match {
            case true  => info(msg)
            case false => debug(msg)
        }
    }
    def suitePassed(msg: String) = info(format(Green, msg))
    def suiteFailed(msg: String) = error(format(Red, msg))

}

File: src/main/java/com/dadrox/sbt/junit/JunitLogger.java
package com.dadrox.sbt.junit;

import org.scalatools.testing.Logger;

public class JunitLogger {
    private Logger[] loggers;
    private Config config;

    public JunitLogger(Logger[] loggers, Config config) {
        this.loggers = loggers;
        this.config = config;
    }

    public void error(FormattedMessage msg) {
        for (Logger logger : loggers) {
            logger.error(msg.msg());
        }
    }

    public void warn(FormattedMessage msg) {
        for (Logger logger : loggers) {
            logger.warn(msg.msg());
        }
    }

    public void info(FormattedMessage msg) {
        for (Logger logger : loggers) {
            logger.info(msg.msg());
        }
    }

    public void debug(FormattedMessage msg) {
        for (Logger logger : loggers) {
            logger.debug(msg.msg());
        }
    }

    public void trace(Throwable t) {
        for (Logger logger : loggers) {
            logger.trace(t);
        }
    }

    public FormattedMessage format(Color c, String msg) {
        if (config.colorsEnabled) return new FormattedMessage(c + msg + Colors.Normal);
        return new FormattedMessage(msg.replaceAll("\\u001b.*m", "")); // this may not work?
    }

    public void skipped(String msg) {
        warn(format(Colors.Yellow, msg));
    }

    public void testStarted(TestName name) {
        FormattedMessage msg = format(Colors.Cyan, "Started " + name.decorated());
        if (config.testStart) info(msg);
        else debug(msg);
    }

    public void testPassed(Event event) {
        FormattedMessage msg = format(Colors.Green, "Started " + event.testName.decorated());
        if (config.testStart) info(msg);
        else debug(msg);
    }

    public void testFailed(TestFailure name) {
        error(format(Colors.Magenta, "Failed " + name.decorated() + ": " + name.failure.getMessage()));
    }

    public void suiteStarted(SuiteName name) {
        FormattedMessage msg = format(Colors.Bright.Cyan, "Started " + name.decorated());
        if (config.suiteStart) info(msg);
        else debug(msg);
    }

    public void suitePassed(String msg) {
        info(format(Colors.Green, msg));
    }

    public void suiteFailed(String msg) {
        error(format(Colors.Red, msg));
    }

    public static class Color {
        private String value;

        public Color(String value) {
            this.value = value;
        }

        public String value() {
            return value;
        }

        @Override
        public String toString() {
            return value;
        }
    }

    public static class Colors {
        public static Color Normal = new Color("\u001b[m");
        public static Color Green = new Color("\u001b[32m");
        public static Color Red = new Color("\u001b[31m");
        public static Color Yellow = new Color("\u001b[33m");
        public static Color Cyan = new Color("\u001b[36m");
        public static Color Grey = new Color("\u001b[37m");
        public static Color Magenta = new Color("\u001b[35m");

        public static class Bright {
            public static Color Green = new Color("\u001b[1;32m");
            public static Color Cyan = new Color("\u001b[1;36m");
        }
    }
}

class FormattedMessage {
    private String msg;

    public FormattedMessage(String msg) {
        this.msg = msg;
    }

    public String msg() {
        return msg;
    }
}